src/daft-plan/src/physical_optimization/rules/reorder_partition_keys.rs

```diff
@@ -0,0 +1,298 @@
+                PhysicalPlan::Coalesce(..) |
+                PhysicalPlan::FanoutRandom(..) |
+                PhysicalPlan::FanoutByRange(..) |
+                PhysicalPlan::Concat(..) |
+                PhysicalPlan::HashJoin(..) |
+                PhysicalPlan::SortMergeJoin(..) |
+                PhysicalPlan::BroadcastJoin(..) => unreachable!("PhysicalPlan match for ReorderPartitionKeys physical optimizer rule should not be reachable"),
+                #[cfg(feature = "python")]
+                PhysicalPlan::IcebergWrite(..) | PhysicalPlan::DeltaLakeWrite(..) | PhysicalPlan::LanceWrite(..) => {
+                    unreachable!("PhysicalPlan match for ReorderPartitionKeys physical optimizer rule should not be reachable")
+                }
+            }
+        })?;
+        res_transformed.map_data(|c| Ok(c.plan))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use common_error::{DaftError, DaftResult};
+    use common_treenode::{TreeNode, TreeNodeRecursion};
+    use daft_core::{
+        datatypes::Field,
+        schema::{Schema, SchemaRef},
+    };
+    use daft_dsl::{col, ExprRef};
+
+    use crate::{
+        partitioning::UnknownClusteringConfig,
+        physical_ops::{EmptyScan, FanoutByHash, HashJoin, ReduceMerge},
+        physical_optimization::{
+            rules::reorder_partition_keys::ReorderPartitionKeys, rules::PhysicalOptimizerRule,
+        },
+        ClusteringSpec, PhysicalPlan, PhysicalPlanRef,
+    };
+
+    fn create_dummy_plan(schema: SchemaRef, num_partitions: usize) -> PhysicalPlanRef {
+        PhysicalPlan::EmptyScan(EmptyScan::new(
+            schema,
+            ClusteringSpec::Unknown(UnknownClusteringConfig::new(num_partitions)).into(),
+        ))
+        .into()
+    }
+
+    fn add_repartition(
+        plan: PhysicalPlanRef,
+        num_partitions: usize,
+        partition_by: Vec<ExprRef>,
+    ) -> PhysicalPlanRef {
+        PhysicalPlan::ReduceMerge(ReduceMerge::new(
+            PhysicalPlan::FanoutByHash(FanoutByHash::new(plan, num_partitions, partition_by))
+                .into(),
+        ))
+        .into()
+    }
+
+    // ensure all repartitions are equal to cols
+    fn check_repartitions(plan: PhysicalPlanRef, cols: Vec<ExprRef>) -> DaftResult<()> {
+        plan.apply(|p| match p.as_ref() {
+            PhysicalPlan::FanoutByHash(FanoutByHash { partition_by, .. }) => {
+                if *partition_by != cols {
+                    Err(DaftError::ComputeError(format!(
+                        "Columns are different: got {:?}",
+                        partition_by
+                    )))
+                } else {
+                    Ok(TreeNodeRecursion::Continue)
+                }
+            }
+            _ => Ok(TreeNodeRecursion::Continue),
+        })
+        .map(|_| ())
+    }
+
+    // makes sure trivial repartitions are modified
+    #[test]
+    fn test_repartition_modified() -> DaftResult<()> {
```

Comment:
not necessary, but I think instead of comparing aspects of the plan, it'd help visualize what the rules are actually doing if we compared the entire plan compared to an expected plan.

something like

```rs
fn test_repartition_modified() -> DaftResult<()> {
    let actual_plan: PhysicalPlanRef = todo!("actual plan");
    let actual_plan = rule.rewrite(actual)?;
    let expected_plan: PhysicalPlanRef = todo!("expected plan");
    assert_eq!(actual, expected);
}
```

Reply:
Added!